### PR TITLE
20171129 forgotpassword screen

### PR DESCRIPTION
**Pull request by at-thinhnguyen2**

### CODE REVIEW

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+//  ForgotPasswordViewController.swift
+//  MyApp
+//
+//  Created by Thinh Nguyen X. on 11/29/17.
+//  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import UIKit
+import MVVM
+class ForgotPasswordViewController: UIViewController, MVVM.View {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: thêm linespace

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    var viewModel = ForgotPasswordViewModel() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: lthêm linespace

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        didSet {
+            updateView()
+        }
+    }
+    @IBOutlet weak var doneButton: UIButton!
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: thêm linespace

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    @IBOutlet weak var backButton: UIButton!
+    @IBOutlet weak var emailTextField: UITextField!
+    var loginVC = LoginViewController()
+    // MARK: - Life cycle
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: thêm linespace

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupUI()
+        setupActions()
+    }
+    func setupUI() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: thêm linespace, private function

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboard))
+        view.addGestureRecognizer(tap)
+        emailTextField.delegate = self
+        emailTextField.returnKeyType = .done
+        emailTextField.keyboardType  = .emailAddress
+    }
+    @objc func dismissKeyboard() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: thêm linespace, private function

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        view.endEditing(true)
+    }
+    func textField(for field: ForgotPasswordViewModel.Field) -> UITextField {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: thêm linespace, private function

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        switch field {
+        case .email: return emailTextField
+        }
+    }
+}
+// MARK - Private
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: line space

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+extension ForgotPasswordViewController {
+    fileprivate func setupActions() {
+        doneButton.addTarget(self, action: #selector(done), for: .touchUpInside)
+        backButton.addTarget(self, action: #selector(back), for: .touchUpInside)
+    }
+    @objc func done() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: fileprivate

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        case.failure(let field, let msg):
+            showAlertError(msg: msg, field: field)
+        }
+    }
+    @objc private func back() {
+        if let navi = navigationController {
+            navi.popViewController(animated: true)
+        }
+    }
+    func backLogin() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: linespace, private?

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+                case .failure(let error):
+                    this.alert(error: error)
+                }
+                this.viewDidUpdated()
+            }
+        case.failure(let field, let msg):
+            showAlertError(msg: msg, field: field)
+        }
+    }
+    @objc private func back() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: linespace

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        alert(title: App.String.forgotPassword, msg: App.String.msgSuccess, buttons: [App.String.cancel, App.String.ok], handler: { [weak self] (actions) in
+            guard let this = self, let title = actions.title else { return }
+            switch title {
+            case App.String.ok:
+                this.popToLogin()
+            default: break
+            }
+        })
+    }
+    func showAlertError(msg: String, field: ForgotPasswordViewModel.Field) {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: linespace, private?

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        alert(title: App.String.error, msg: msg, buttons: [App.String.ok], handler: { [weak self] _ in
+            guard let this = self else { return }
+            let fld = this.textField(for: field)
+            fld.becomeFirstResponder()
+        })
+        viewDidUpdated()
+    }
+    func updateView() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: linespace, private?

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        guard isViewLoaded else { return }
+        emailTextField.text = viewModel.email
+    }
+    func popToLogin() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: linespace, private?

File: MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        let navi = UINavigationController(rootViewController: loginVC)
+        AppDelegate.shared.window?.rootViewController = navi
+    }
+}
+// MARK - TextField Delegate
```
Comment (at-linhvo) [inline MyApp/View/Controller/Forgot/ForgotPasswordViewController.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+//  MyApp
+//
+//  Created by Thinh Nguyen X. on 11/29/17.
+//  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import Foundation
+import SwiftUtils
+import MVVM
+final class ForgotPasswordViewModel: MVVM.ViewModel {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    enum Field: String {
+        case email
+    }
+    enum Validation: CustomStringConvertible {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+        case success
+        case failure(field: Field, msg: String)
+        var isSuccess: Bool {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            switch self {
+            case .success: return true
+            default: return false
+            }
+        }
+        var description: String {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+            switch self {
+            case .success: return App.String.success
+            case .failure(_, let msg):
+                return App.String.failure + msg
+            }
+        }
+    }
+    var email = ""
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    func validate() -> Validation {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+        guard email.len != 0 else {
+            return .failure(field: .email, msg: App.String.emailEmpty)
+        }
+        guard email.validate(String.Regex.Email1) else {
+            return .failure(field: .email, msg: App.String.emailFormat)
+        }
+        return .success
+    }
+    enum ForgotResult {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: đưa lên trên

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+        case success
+        case failure(Error)
+    }
+    func forgot(_ completion: @escaping (ForgotResult) -> Void) {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+            return
+        }
+        guard loadPlist(email: email) else {
+            let error = NSError(message: "Error")
+            completion(.failure(error))
+            return
+        }
+        completion(.success)
+    }
+    func loadPlist(email: String) -> Bool {
```
Comment (at-linhvo) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: linespace

File: MyApp/View/Controller/Login/LoginViewController.swift [hunk 25]
```diff
@@ -63,14 +65,15 @@ final class LoginViewController: ViewController {
...
+        forgotButton.addTarget(self, action: #selector(forgot), for: .touchUpInside)
     }
 
     @objc fileprivate func login() {
         viewModel.username = usernameTextField.string.trimmed
         viewModel.password = passwordTextField.string.trimmed
         switch viewModel.validate() {
         case .success:
-            viewModel.login { [weak self] (result) in
+            viewModel.login { [weak self ](result) in
```
Comment (at-linhvo) [inline MyApp/View/Controller/Login/LoginViewController.swift]: [weak self ] thừa dấu space

File: MyApp/ViewModel/ForgotPasswordViewModel.swift [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+    enum Validation: CustomStringConvertible {
+        case success
+        case failure(field: Field, msg: String)
+        var isSuccess: Bool {
+            switch self {
+            case .success: return true
+            default: return false
+            }
+        }
+
```
Comment (at-daiho) [inline MyApp/ViewModel/ForgotPasswordViewModel.swift]: WARNING - Lines should not have trailing whitespace.

**Comment by at-daiho** [comment]
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

**Pull request merged**